# Genetic Data in R and GDS

## Genetic Data in R and GDS lecture

Here's a recording of this lecture (50 minutes 56 seconds):


[Recording](https://pitt.hosted.panopto.com/Panopto/Pages/Viewer.aspx?id=bda73d86-08f7-4159-876c-b2310136c45b){.external target="_blank"}

<iframe credentialless src="https://pitt.hosted.panopto.com/Panopto/Pages/Embed.aspx?id=bda73d86-08f7-4159-876c-b2310136c45b&amp;autoplay=false&amp;offerviewer=true&amp;showtitle=true&amp;showbrand=true&amp;captions=true&amp;interactivity=all" height="405" width="720" style="border: 1px solid #464646;" allowfullscreen allow="autoplay">

</iframe>

## Genetic Data in R and GDS slides

[PDF slide set](pdfs/genetic_data_in_r_gds_lecture.pdf){target="_blank"}

<iframe width="100%" height="800" src="pdfs/genetic_data_in_r_gds_lecture.pdf">

</iframe>

## Questions

### Question 1

Which of the following are true? Select the correct options (more than one may be correct).

```{r, echo=FALSE}
library(webexercises)
choices = c(
"seqArray GDS and SNP GDS files are completely interchangeable, and all functions that work with one format will work with the other",
"they are not interchangeable, but they can be easily converted into each other with plink",
answer = "They are not interchangeable, but they can be easily converted into each other with the seqArray package.",
answer = "plink files can be converted into GDS files using the seqArray package"
)
```

`r longmcq(choices)`


### Question 2

How do you open, close, and save GDS files in R?

```{r, echo=FALSE}
library(webexercises)
choices = c(
"read.table and write.table",
"fread and fwrite",
"BEDMatrix",
"seqGDS2VCF",
answer="seqOpen, seqClose, and seqExport"
)
```

`r longmcq(choices)`

### Question 3

How do you access data in a GDS object after opening it?

```{r, echo=FALSE}
library(webexercises)
choices = c(
"the dollar-sign operator",
"piping",
"redirection",
"square brackets",
answer="seqGetData and other built-in accessor functions",
"double square brackets"
)
```

`r longmcq(choices)`

### Question 4

How would you filter a GDS object and then save the filtered data to a new GDS file?

```{r, echo=FALSE}
library(webexercises)
choices = c(
"seqGetFilter and then seqExport",
"seqSetFilter and then seqClose",
answer="seqSetFilter and then seqExport",
"seqSetFilter and then write.table"
)
```

`r longmcq(choices)`

### Question 5

Which of the following are true? Select all that apply.

```{r, echo=FALSE}
library(webexercises)
choices = c(
"GDS files cannot store the same information as a VCF file",
answer="GDS files are more flexible than plink file formats",
answer="GDS files are smaller and more efficient than plink and VCF files",
answer="GDS files integrate with specialized packages for GWAS data cleaning, kinship and population structure calculations, and association analysis",
"When GDS files are opened, the entire dataset is read into memory",
"seqArray cannot be used to convert from VCF to GDS and from GDS to VCF",
answer="plink can be used to convert from plink to VCF and from VCF to plink"
)
```

`r longmcq(choices)`